### PR TITLE
Fix GridFunction::GetGradient for face neighbor elements

### LINKED ISSUE #1801: Maybe a bug in ParBilinearForm::AssembleSharedFaces()
Hi all,

When I run the below code in serial, it's OK. But it will output below error info in parallel.
```
Verification failed: (i < mesh->GetNE()) is false:
 --> Invalid element id 868, maximum allowed 769
 ... in function: const mfem::FiniteElement* mfem::FiniteElementSpace::GetFE(int) const
 ... in file: /home/fan/mfem/fem/fespace.cpp:1801

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
```
Through debugging, I find there may be a bug in `ParBilinearForm::AssembleSharedFaces()`.

The whole source code (MWE) is below
```c++
#include <iostream>
#include "mfem.hpp"

using namespace std;
using namespace mfem;


/* Integral on a face:
 *
 *     q <{u grad(w).n}, [v]>_E,
 *
 * u is Trial function, v is Test function
 * q is given Coefficient
 * w is GridFunction
 * */
class DGEdgeIntegrator1: public BilinearFormIntegrator
{
protected:
    Coefficient *Q;
    ParGridFunction& w;
    GradientGridFunctionCoefficient* gradw;

    Vector nor, shape1, shape2, grad_w;
    double val1, val2;

public:
    DGEdgeIntegrator1(ParGridFunction &w_) : Q(NULL), w(w_)
    { gradw = new GradientGridFunctionCoefficient(&w); }

    DGEdgeIntegrator1(Coefficient &q, ParGridFunction &w_) : Q(&q), w(w_)
    { gradw = new GradientGridFunctionCoefficient(&w); }

    ~DGEdgeIntegrator1() { delete gradw; }

    using BilinearFormIntegrator::AssembleFaceMatrix;
    virtual void AssembleFaceMatrix(const FiniteElement &el1,
                                    const FiniteElement &el2,
                                    FaceElementTransformations &Trans,
                                    DenseMatrix &elmat)
    {
        int dim, ndof1(0), ndof2(0);

        dim = el1.GetDim();

        grad_w.SetSize(dim);
        nor.SetSize(dim);

        ndof1 = el1.GetDof();
        if (Trans.Elem2No >= 0) // interior boundary
        {
            ndof2 = el2.GetDof();
        }

        shape1.SetSize(ndof1);
        shape2.SetSize(ndof2);

        elmat.SetSize(ndof1 + ndof2);
        elmat = 0.0;

        const IntegrationRule *ir = IntRule; // ref: DGTraceIntegrator::AssembleFaceMatrix
        if (ir == NULL) {
            int order;
            // Assuming order(u)==order(mesh)
            if (ndof2)
                order = (min(Trans.Elem1->OrderW(), Trans.Elem2->OrderW()) +
                         2 * max(el1.GetOrder(), el2.GetOrder()));
            else {
                order = Trans.Elem1->OrderW() + 2 * el1.GetOrder();
            }
            if (el1.Space() == FunctionSpace::Pk) {
                order++;
            }
            ir = &IntRules.Get(Trans.FaceGeom, order); //得到face上的积分规则(里面包含积分点)
        }

        for (int p=0; p<ir->GetNPoints(); ++p)
        {
            const IntegrationPoint& ip = ir->IntPoint(p);

            Trans.SetAllIntPoints(&ip);

            const IntegrationPoint &eip1 = Trans.GetElement1IntPoint();
            const IntegrationPoint &eip2 = Trans.GetElement2IntPoint();

            el1.CalcShape(eip1, shape1);
            el2.CalcShape(eip2, shape2);

            if (dim == 1) {
                nor(0) = 2 * eip1.x - 1.0;
            }
            else {
                CalcOrtho(Trans.Jacobian(), nor); // 计算Face的法向量
            }

            gradw->Eval(grad_w, *Trans.Elem1, eip1); // OK: parallel and serial
            val1 = ip.weight * Q->Eval(*Trans.Elem1, eip1) * (grad_w * nor);

            if (Trans.Elem2No >= 0) // interior boundary
            {
                val1 *= 0.5;
                for (int i=0; i<ndof1; ++i)
                    for (int j=0; j<ndof1; ++j)
                        elmat(i, j) += val1 * shape1(i) * shape1(j);

                if (Trans.Elem2No >= w.FESpace()->GetNE())
                {
                    cout << "Elem2NO: " << Trans.Elem2No << ", " << w.FESpace()->GetNE() << ", " << w.ParFESpace()->GetNE() << endl;
                }
                gradw->Eval(grad_w, *Trans.Elem2, eip2); // Error: parallel; OK: serial
//                w.GetGradient(*Trans.Elem2, grad_w); // same as above

                val2 = ip.weight * Q->Eval(*Trans.Elem2, eip2) * (grad_w * nor) * 0.5;

                for (int i=0; i<ndof2; ++i)
                    for (int j=0; j<ndof1; ++j)
                        elmat(ndof1 + i, j) -= shape2(i) * val1 * shape1(j);

                for (int i=0; i<ndof1; ++i)
                    for (int j=0; j<ndof2; ++j)
                        elmat(i, ndof1+j) += shape1(i) * val2 * shape2(j);

                for (int i=0; i<ndof2; ++i)
                    for (int j=0; j<ndof2; ++j)
                        elmat(ndof1 + i, ndof1 + j) -= val2 * shape2(i) * shape2(j);
            }
            else
            {
                for (int i=0; i<ndof1; ++i)
                    for (int j=0; j<ndof1; ++j)
                        elmat(i, j) += val1 * shape1(i) * shape1(j);
            }
        }
    }
};

int main(int argc, char *argv[])
{
    MPI_Init(&argc, &argv);

    Mesh* mesh = new Mesh(8, 8, 8, Element::TETRAHEDRON);
    ParMesh* pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
    delete mesh;

    DG_FECollection* fec = new DG_FECollection(1, 3);
    ParFiniteElementSpace* fes = new ParFiniteElementSpace(pmesh, fec);

    ParGridFunction* phi_gf = new ParGridFunction(fes);
    *phi_gf = 0.0;

    ParBilinearForm* e1 = new ParBilinearForm(fes);
    ConstantCoefficient one(1.0);
    e1->AddInteriorFaceIntegrator(new DGEdgeIntegrator1(one, *phi_gf)); // error for parallel, ok for serial
//    e1->AddBdrFaceIntegrator(new DGEdgeIntegrator1(one, *phi_gf));      // ok for serial and parallel
    e1->Assemble(0);

    delete e1;
    MPI_Finalize();
}
```

### PR DESCRIPTION
**Pull request by pazner**
Fixes #1801 (I think), and simultaneously attempts to make some of the face neighbor handling more general/automatic.

I'm not sure this approach is the best one, but I do believe that with the correct abstraction, the face neighbor handling can be made much easier.

Ideally, users/callers shouldn't care if the referenced element is a "local element" or "face neighbor element" (or, potentially in the future, a vertex neighbor, element in the ghost layer, etc.), the standard MFEM functionality should work transparently. I think this means that the "low-level" `Mesh` and `FiniteElementSpace` calls should take care of distinguishing between these element classes, and the higher-level functions can be built on the low-level primitives.

I'm not thrilled that this approach makes member functions like `FiniteElementSpace::GetFE` virtual. Perhaps this can be avoided somehow.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1829](https://github.com/mfem/mfem/pull/1829) | @pazner | @v-dobrev | @mlstowell + @jakubcerveny | 10/29/20 | 12/22/20 | 12/30/20 | |

### CODE REVIEW

**Comment by pazner** [comment]
> This looks good but I do have a couple concerns.
> 
> There are still a handful of methods which take element indices and will probably fail if face neighbor element indices are used. It's beyond the scope of this PR to generalize all of them but we should at least add comments to them which point out this limitation. The methods I have in mind range from `FiniteElementSpace::GetElementType` to `FiniteElementSpace::GetAttribute` and `FiniteElementSpace::GetElementInteriorDofs`. There are probably others as well.
> 
> The other matter is the need for a unit test. Perhaps a modified (and much shorter) version of `tests/unit/fem/test_get_value.cpp` would be appropriate here.

These are good points, thanks @mlstowell. I will address them.

**Comment by v-dobrev** [comment]
@jakubcerveny, can you review this PR?

File: fem/gridfunc.cpp [hunk 1]
```diff
@@ -1550,18 +1550,16 @@ void GridFunction::GetGradient(ElementTransformation &T, Vector &grad) const
...
          int spaceDim = fes->GetMesh()->SpaceDimension();
          int dim = fe->GetDim(), dof = fe->GetDof();
          DenseMatrix dshape(dof, dim);
          Vector lval, gh(dim);
-         Array<int> dofs;
 
          grad.SetSize(spaceDim);
-         fes->GetElementDofs(T.ElementNo, dofs);
-         GetSubVector(dofs, lval);
+         GetElementDofs(T.ElementNo, lval);
```
Comment (jakubcerveny) [inline fem/gridfunc.cpp:1562]: This confused me, I suggest we call it `GridFunction::GetElementDofValues`

File: fem/fespace.hpp [hunk 2]
```diff
@@ -571,7 +571,7 @@ class FiniteElementSpace
 
    /** @brief Returns pointer to the FiniteElement in the FiniteElementCollection
         associated with i'th element in the mesh object. */
-   const FiniteElement *GetFE(int i) const;
+   virtual const FiniteElement *GetFE(int i) const;
```
Comment (jakubcerveny) [inline fem/fespace.hpp:574]: I think we don't have to touch anything in the `(Par)FiniteElementSpace` in this PR. Just like with `GridFunction::GetElementDofValues`, we can have `GridFunction::GetFE` that does the hiding of the difference between regular/neighbor elements.

**Review by jakubcerveny — CHANGES_REQUESTED** [review]
Sorry for the late review...

I think this is OK, but I would make a few changes. I think we can keep the unification of regular/neighbor elements at the (Par)GridFunction level, at least for the time being (in the future, we could perhaps think about unifying on the FiniteElementSpace level, but that would be a bigger overhaul).

I suggest:
- create a (probably internal) virtual method (Par)GridFunction::GetFE that handles regular/neighbor uniformly
- create/rename virtual (Par)GridFunction::GetElementDofValues that retrieves regular/neighbor values uniformly

This way it can all stay at the (Par)GridFunction level, hopefully? Also no need to make FiniteElementSpace::GetFE virtual.

File: fem/gridfunc.cpp [hunk 3]
```diff
@@ -1731,6 +1729,13 @@ void GridFunction::GetElementAverages(GridFunction &avgs) const
    }
 }
 
+void GridFunction::GetElementDofs(int el, Vector &dof_vals) const
+{
+   Array<int> dof_idx;
+   fes->GetElementVDofs(el, dof_idx);
```
Comment (jakubcerveny) [inline fem/gridfunc.cpp:1735]: This is different than the original (which calls `GetElementDofs`), is that intentional?

File: fem/gridfunc.cpp (hunk 1 above)
Comment (pazner) [reply fem/gridfunc.cpp]: I agree, this name is better. I changed it.

File: fem/gridfunc.cpp (hunk 3 above)
Comment (pazner) [reply fem/gridfunc.cpp:1735]: It seems to be that vdofs is almost always the more useful result. I'm not sure I see the use-case here for `GetElementDofs` for a non-scalar grid function (for scalar they should do the same thing).

**Comment by pazner** [comment]
> I think this is OK, but I would make a few changes. I think we can keep the unification of regular/neighbor elements at the (Par)GridFunction level, at least for the time being (in the future, we could perhaps think about unifying on the FiniteElementSpace level, but that would be a bigger overhaul).
> 
> I suggest:
> 
> * create a (probably internal) virtual method (Par)GridFunction::GetFE that handles regular/neighbor uniformly
> * create/rename virtual (Par)GridFunction::GetElementDofValues that retrieves regular/neighbor values uniformly
> 
> This way it can all stay at the (Par)GridFunction level, hopefully? Also no need to make FiniteElementSpace::GetFE virtual.

Yes, I think your approach will work, but I am not convinced that it is better. It has the advantage that `FiniteElementSpace::GetFE` does not need to be virtual (so perhaps more efficient, but some benchmarking would be required to see if this is relevant — I doubt that `GetFE` is ever a bottleneck).

I don't think that `GridFunction::GetFE` makes sense conceptually. What does getting the FE of a grid function mean, other than getting the FE of the associated space? And why would the behaviors be different (in parallel only) in a subtle way that is not closely related to the distinction between a GridFunction and a FiniteElementSpace? This seems like an invitation to write buggy code that appears to work properly in serial or for H1 spaces, but will fail in strange ways for DG in parallel as soon as face neighbors are involved.

Perhaps more importantly, `GridFunction::GetGradient` is just one example of a *caller* of `FiniteElementSpace::GetFE` that could benefit from functioning properly for face neighbor elements. There are potentially many other examples of such functions that could be fixed by this change. To me, it makes more sense to be able to treat face neighbor elements as normal elements as much as possible, and therefore reduce the number of special cases, so that as much code will work *automatically* in these cases. The other approach would fix this specific issue, but the same logic would have to be duplicated any other time face neighbors are involved.

I agree that completely unifying on the FiniteElementSpace level would be a (much) bigger overhaul (and I believe it is something we should think about doing), but this small change shouldn't cause other disruptions, so I don't believe it's "all or nothing".

**Comment by pazner** [comment]
With all that being said, I could definitely be convinced that `ParFiniteElementSpace::GetElementVDofs` should fall back on `ParFiniteElementSpace::GetFaceNbrElementVDofs` for face neighbor indices, in which case `GridFunction::GetElementDofValues` would not need a special case face neighbors either :)

**Comment by jakubcerveny** [comment]
My intuition was to prevent handling face neighbors on *both* the GridFunction level and the FiniteElementSpace level. But thinking more about this, it probably cannot be avoided, since ParGridFunction needs to work with the neighbor data. You are probably right.

**Review by jakubcerveny — APPROVED** [review]

**Review by mlstowell — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**